cpp/src/io/json/experimental/read_json.cpp

```diff
@@ -59,16 +70,35 @@ rmm::device_uvector<char> ingest_raw_input(host_span<std::unique_ptr<datasource>
         auto destination = reinterpret_cast<uint8_t*>(d_buffer.data()) + bytes_read;
         if (source->is_device_read_preferred(data_size)) {
           bytes_read += source->device_read(range_offset, data_size, destination, stream);
+          delimiter_map.push_back(bytes_read);
+          bytes_read += num_delimiter_chars;
         } else {
           h_buffers.emplace_back(source->host_read(range_offset, data_size));
           auto const& h_buffer = h_buffers.back();
           CUDF_CUDA_TRY(cudaMemcpyAsync(
             destination, h_buffer->data(), h_buffer->size(), cudaMemcpyDefault, stream.value()));
           bytes_read += h_buffer->size();
+          delimiter_map.push_back(bytes_read);
+          bytes_read += num_delimiter_chars;
         }
       }
     }
 
+    // If this is a multi-file source, we scatter the JSON line delimiters between files
+    if (sources.size() > 1) {
+      static_assert(num_delimiter_chars == 1,
+                    "Currently only single-character delimiters are supported");
+      auto const delimiter_source = thrust::make_constant_iterator('\n');
+      auto const d_delimiter_map  = cudf::detail::make_device_uvector_async(
+        host_span<size_type const>{delimiter_map.data(), delimiter_map.size() - 1},
```

Comment:
> Or is this deliberate because you want to interleave newlines between the sources, but not add a newline to the very last source?

Exactly, that's the idea. 

> (In which case I think the scatter end iterator wants to be .size() - 1?

Yup. Please note, though, that `scatter` is using the `d_delimiter_map.size()`, which already has the proper size (i.e., it already excludes the offset for the last file).

Reply:
> Yup. Please note, though, that `scatter` is using the `d_delimiter_map.size()`, which already has the proper size (i.e., it already excludes the offset for the last file).

Ah missed that, thanks.